applications/call_inspector/src/parsers/ci_chunk.erl

```diff
@@ -0,0 +1,308 @@
+        ]).
+
+-record(ci_chunk, {call_id :: ne_binary()
+                  ,data = [] :: ne_binaries()
+                  ,timestamp :: number()
+                  ,ref_timestamp :: number()
+                  ,src_ip :: ne_binary()
+                  ,src_port :: pos_integer()
+                  ,dst_ip :: ne_binary()
+                  ,dst_port :: pos_integer()
+                  ,parser :: atom()
+                  ,label :: ne_binary()
+                 }).
+-type chunk() :: #ci_chunk{}.
+
+-export_type([chunk/0]).
+
+-include("../call_inspector.hrl").
+
+-define(SETTER(Field),
+        Field(#ci_chunk{}=Chunk, Value) ->
+               Chunk#ci_chunk{Field = Value}
+       ).
+
+-define(GETTER(Field),
+        Field(#ci_chunk{Field = Value}) ->
+               Value
+       ).
+
+
+%% API
+
+-spec new() -> chunk().
+new() -> #ci_chunk{}.
+
+-spec setters(chunk(), [{fun((chunk(), A) -> chunk()), A}]) -> chunk().
+setters(#ci_chunk{}=Chunk, Setters) ->
+    Apply = fun ({Fun, Arg}, C) -> Fun(C, Arg) end,
+    NewChunk = lists:foldl(Apply, Chunk, Setters),
+    NewChunk#ci_chunk{ref_timestamp = ci_parsers_util:timestamp()}.
+
+-spec call_id(chunk(), ne_binary()) -> chunk().
+?SETTER(call_id).
+-spec call_id(chunk()) -> api_binary().
+?GETTER(call_id).
+
+-spec data(chunk(), ne_binaries()) -> chunk().
+?SETTER(data).
+-spec data(chunk()) -> ne_binaries().
+?GETTER(data).
+-spec append_data(chunk(), ne_binary()) -> chunk().
+append_data(#ci_chunk{data=D}=Chunk, Data) ->
+    Chunk#ci_chunk{data=[Data|D]}.
+
+-spec timestamp(chunk(), integer()) -> chunk().
+?SETTER(timestamp).
+-spec timestamp(chunk()) -> api_integer().
+?GETTER(timestamp).
+
+-spec ref_timestamp(chunk()) -> api_number().
+?GETTER(ref_timestamp).
+
+-spec src_ip(chunk(), ne_binary()) -> chunk().
+src_ip(#ci_chunk{}=Chunk, Val) ->
+    Chunk#ci_chunk{src_ip = resolve(Val)}.
+-spec src_ip(chunk()) -> api_binary().
+?GETTER(src_ip).
+
+-spec src_port(chunk(), pos_integer()) -> chunk().
+?SETTER(src_port).
+-spec src_port(chunk()) -> api_pos_integer().
+?GETTER(src_port).
+
+-spec dst_ip(chunk(), ne_binary()) -> chunk().
+dst_ip(#ci_chunk{}=Chunk, Val) ->
+    Chunk#ci_chunk{dst_ip = resolve(Val)}.
+-spec dst_ip(chunk()) -> api_binary().
+?GETTER(dst_ip).
+
+-spec dst_port(chunk(), pos_integer()) -> chunk().
+?SETTER(dst_port).
+-spec dst_port(chunk()) -> api_pos_integer().
+?GETTER(dst_port).
+
+-spec parser(chunk(), atom()) -> chunk().
+?SETTER(parser).
+-spec parser(chunk()) -> api_atom().
+?GETTER(parser).
+
+-spec label(chunk(), ne_binary()) -> chunk().
+?SETTER(label).
+-spec label(chunk()) -> api_binary().
+?GETTER(label).
+
+-spec to_json(chunk()) -> wh_json:object().
+to_json(Chunk) ->
+    wh_json:from_list(
+      [{<<"call-id">>, call_id(Chunk)}
+       ,{<<"timestamp">>, timestamp(Chunk)}
+       ,{<<"ref_timestamp">>, wh_util:to_binary(ref_timestamp(Chunk))}
+       ,{<<"label">>, label(Chunk)}
+       ,{<<"raw">>, data(Chunk)}
+       ,{<<"src">>, src(Chunk)}
```

Comment:
`src/1/` and `dst/1` appear to return 2-tuples. `{<<"src">>, {Host, Port}}` is not a valid JSON term. Should either be converted to a JSON object, combined into a binary, or something that is valid.